File: CPPSolutions/CPPSolutions/M817_链表组件.cpp
#include <cstdio>
#include <unordered_set>
#include "ListNode.h"

using namespace std;

class Solution {
public:
    int numComponents(ListNode* head, vector<int>& nums) {
		unordered_set<int> set;
		for (int num : nums)
			set.emplace(num);

		ListNode* node = head;
		int ans = 0;
		bool flag = false;
		while (node != nullptr) {
			if (set.count(node->val)) {
				if (!flag) {
					flag = true;
					ans++;
				} 
			}
			else {
				flag = false;
			}
			node = node->next;
		}

		return ans;
	}
};

File: GolangSolutions/LinkedList/M817_链表组件.go
package LinkedList

type ListNode struct {
	Val  int
	Next *ListNode
}

// 可以直接在(ans int)声明返回值，返回时直接写return，ans其作用域依然是函数本身
func numComponents(head *ListNode, nums []int) (ans int) {
	set := make(map[int]struct{}, len(nums))
	for _, v := range nums { // range返回的是[index, value], _为忽略前一个返回值
		set[v] = struct{}{}
	}

	node := head
	for flag := false; node != nil; node = node.Next {
		if _, ok := set[node.Val]; !ok {
			flag = false
		} else if !flag {
			flag = true
			ans++
		}
	}

	return
}
